toqito/measurement_ops/measure.py

```diff
@@ -1,76 +1,118 @@
 
-    .. math::
-        \langle P_1, \rho \rangle = \frac{2}{3}.
+    .. jupyter-execute::
 
-    >>> measure(proj_1, rho)
-    0.6666666666666666
+        import numpy as np
+        from toqito.measurement_ops.measure import measure
+        rho = np.array([[0.5, 0.5], [0.5, 0.5]])
+        K0 = np.array([[1, 0], [0, 0]])
+        K1 = np.array([[0, 0], [0, 1]])
 
-    References
-    ==========
-    .. bibliography::
-        :filter: docname in docnames
+        # Returns list of probabilities.
+        print(measure(rho, [K0, K1]))
 
+        # Returns list of (probability, post_state) tuples.
+        print(measure(rho, [K0, K1], update=True))
 
-    :param measurement: The measurement to apply.
-    :param state: The state to apply the measurement to.
-    :return: Returns the probability of obtaining a given outcome after applying
-             the variable :code:`measurement` to the variable :code:`state`.
+    :param state: Quantum state as a density matrix (an np.ndarray with shape (d, d)).
+    :param measurement: Either a single measurement operator (an np.ndarray) or a list/tuple of operators.
+                        When providing a list, they are assumed to be Kraus operators satisfying the completeness
+                        relation.
+    :param tol: Tolerance for numerical precision (default is 1e-10).
+    :param update: If True, also return the post-measurement state(s); otherwise, only the probability or
+                   probabilities are returned.
+    :raises ValueError: If a list of operators does not satisfy the completeness relation.
+    :return: If a single operator is provided, returns a float (probability) or a tuple (probability, post_state)
+             if `update` is True. If a list is provided, returns a list of probabilities or a list of tuples if
+             `update` is True.
 
     """
-    return float(np.trace(measurement.conj().T @ state))
+    # Single-operator case
+    if not isinstance(measurement, (list, tuple)):
+        result = measurement @ state @ measurement.conj().T
+        prob = np.trace(result).real
+        if prob > tol:
+            post_state = result / prob
+        else:
+            post_state = np.zeros_like(state)
+        return (prob, post_state) if update else prob
+
+    # List-of-operators case
+    outcomes: list[float | tuple[float, np.ndarray]] = []
+    probs: list[float] = []
+
+    for op in measurement:
+        result = op @ state @ op.conj().T
+        prob = np.trace(result).real
+        probs.append(prob)
+
+        if prob > tol:
+            post_state = result / prob
+        else:
+            post_state = np.zeros_like(state)
+
+        outcomes.append((prob, post_state) if update else prob)
+
+    # Only enforce completeness if we're doing the update AND every outcome was nonzero
+    if update and all(p > tol for p in probs):
+        d = state.shape[0]
+        completeness = sum(op.T.conj() @ op for op in measurement)
```

Comment:
I think the completeness relation should be verified before L99. 

Maybe I am missing some condition as to why you check this only if L112 is satisfied?

Reply:
It's a performance compromise. This defers the completeness check until after measurement application, and only checks it if all outcome probabilities are nonzero.